src/passes/OptimizeInstructions.cpp

```diff
@@ -1862,6 +1862,191 @@ struct OptimizeInstructions
     }
   }
 
+  void visitStructRMW(StructRMW* curr) {
+    skipNonNullCast(curr->ref, curr);
+    if (trapOnNull(curr, curr->ref)) {
+      return;
+    }
+
+    if (!curr->ref->type.isStruct()) {
+      return;
+    }
+
+    Builder builder(*getModule());
+
+    // Even when the access to shared memory, we can optimize out the modify and
+    // write parts if we know that the modified value is the same as the
+    // original value. This is valid because reads from writes that don't change
+    // the in-memory value can be considered to be reads from the previous write
+    // to the same location instead. That means there is no read that
+    // necessarily synchronizes with the write.
+    auto* value =
+      Properties::getFallthrough(curr->value, getPassOptions(), *getModule());
+    if (Properties::isSingleConstantExpression(value)) {
+      auto val = Properties::getLiteral(value);
+      bool canOptimize = false;
+      switch (curr->op) {
+        case RMWAdd:
+        case RMWSub:
+        case RMWOr:
+        case RMWXor:
+          canOptimize = val.getInteger() == 0;
+          break;
+        case RMWAnd:
+          canOptimize = val == Literal::makeNegOne(val.type);
+          break;
+        case RMWXchg:
+          canOptimize = false;
+          break;
+      }
+      if (canOptimize) {
+        replaceCurrent(builder.makeStructGet(
+          curr->index,
+          getResultOfFirst(curr->ref, builder.makeDrop(curr->value)),
+          curr->order,
+          curr->type));
+        return;
+      }
+    }
+
+    if (curr->ref->type.getHeapType().isShared()) {
+      return;
+    }
+
+    // Lower the RMW to its more basic operations. Breaking the atomic
+    // operation into several non-atomic operations is safe because no other
+    // thread can observe an intermediate state in the unshared memory. This
+    // initially increases code size, but the more basic operations may be
+    // more optimizable than the original RMW.
```

Comment:
How likely are we to succeed? I am somewhat worried that generally the size increase here will stick around, and might also be slower.

Reply:
I'm not sure! We can experiment with it once we have real test cases. I figure it will be easier to experimentally disable this optimization than it would be to write the optimization just for an experiment, though. I could imagine in the long run we would want to do this only when optimizing for speed over size.